File: common/src/main/java/com/kt/kol/common/model/PageDTO.java
package com.kt.kol.common.model;

import java.util.List;

public record PageDTO<T>(
    List<T> content,
    int page,
    int size,
    long totalElements,
    int totalPages,
    boolean hasNext,
    boolean hasPrevious
) {
    public static <T> PageDTO<T> of(List<T> content, int page, int size, long totalElements) {
        int totalPages = (int) Math.ceil((double) totalElements / size);
        boolean hasNext = page < totalPages - 1;
        boolean hasPrevious = page > 0;

        return new PageDTO<>(content, page, size, totalElements, totalPages, hasNext, hasPrevious);
    }

    public static <T> PageDTO<T> empty(int page, int size) {
        return new PageDTO<>(List.of(), page, size, 0L, 0, false, false);
    }
}

File: common/src/main/kotlin/com/kt/kol/common/model/PageDTO.kt
package com.kt.kol.common.model

import kotlin.math.ceil

data class PageDTO<T>(
    val content: List<T>,
    val page: Int,
    val size: Int,
    val totalElements: Long,
    val totalPages: Int,
    val hasNext: Boolean,
    val hasPrevious: Boolean
) {
    companion object {
        fun <T> of(content: List<T>, page: Int, size: Int, totalElements: Long): PageDTO<T> {
            val totalPages = ceil(totalElements.toDouble() / size).toInt()
            val hasNext = page < totalPages - 1
            val hasPrevious = page > 0

            return PageDTO(content, page, size, totalElements, totalPages, hasNext, hasPrevious)
        }

        fun <T> empty(page: Int, size: Int): PageDTO<T> = PageDTO(
            emptyList(), page, size, 0L, 0, false, false
        )
    }
}